File: 0146-lru-cache/0146-lru-cache.js
/**
 * @param {number} capacity
 */
var LRUCache = function(capacity) {
    this.max = capacity;
    this.cache = new Map();
};

/** 
 * @param {number} key
 * @return {number}
 */
LRUCache.prototype.get = function(key) {
    let item = this.cache.get(key);
        
    if(item !== undefined) {
        this.cache.delete(key);
        this.cache.set(key, item);
    } else {
        return -1;
    }
    
    return item;
};

/** 
 * @param {number} key 
 * @param {number} value
 * @return {void}
 */
LRUCache.prototype.put = function(key, value) {
    
    // if the key is already present delete it
    if(this.cache.has(key)) this.cache.delete(key);
    // before inserting the new key check the size of the 
    else if(this.cache.size === this.max) {
    // here the least recently used key will be at top hence we are deleting the first key
        const firstKey = this.cache.keys().next().value;
            
        this.cache.delete(firstKey);
    }
    
    // insert the new key 
    this.cache.set(key, value);
    
    return null;
};

File: 0146-lru-cache/0146-lru-cache.go
// since struct are pass by value hence we need to use pointer
type Node struct {
    key, val int
    prev, next *Node
}

// left and right will be treat as two extreme ends 
// left next will always be least recently used
// right prev will always be most recently used
type LRUCache struct {
    cap         int
    cache       map[int]*Node
    left, right *Node
}


func Constructor(capacity int) LRUCache {
    lru := LRUCache{
        cache: make(map[int]*Node),
        cap: capacity,
        left: &Node{},
        right: &Node{},
    }
    
    lru.left.next = lru.right
    lru.right.prev = lru.left
    return lru
}

func (lru *LRUCache) remove(node *Node) {
    prev, next := node.prev, node.next
    prev.next = next
    next.prev = prev
}

func (lru *LRUCache) insert(node *Node) {
    prev, next := lru.right.prev, lru.right
    // insert between prev and next
    prev.next = node
    next.prev = node
    node.next = next
    node.prev = prev
}


func (lru *LRUCache) Get(key int) int {
    cache := lru.cache
    if node, ok := cache[key]; ok {
        // update location make it most recent
        lru.remove(node)
        lru.insert(node)
        return node.val
    } 
    
    return -1 
}


func (lru *LRUCache) Put(key int, val int)  {
    left := lru.left
    cache := lru.cache
    cap := lru.cap
    // check if the element already exists
    if node, ok := cache[key]; ok {
        // remove this element
        delete(cache, key)
        lru.remove(node)
    }
    
    node := &Node{key: key, val: val}
    cache[key] = node
    lru.insert(node)
    
    if len(cache) > cap {
        // remove least recently used
        delete(cache, left.next.key)
        lru.remove(left.next)

    }
} 


/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */